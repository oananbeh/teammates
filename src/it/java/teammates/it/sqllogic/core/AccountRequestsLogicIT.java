package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.UUID;

import org.testng.annotations.Test;

import teammates.common.datatransfer.AccountRequestStatus;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountRequestsLogic;
import teammates.storage.sqlapi.AccountRequestsDb;
import teammates.storage.sqlentity.AccountRequest;

/**
 * SUT: {@link AccountRequestsLogic}.
 */
public class AccountRequestsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountRequestsLogic accountRequestsLogic = AccountRequestsLogic.inst();

    @Test
    public void testGetAccountRequest_nonExistentAccountRequest_returnsNull() {
        UUID id = UUID.randomUUID();
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertNull(actualAccountRequest);
    }

    @Test
    public void testGetAccountRequest_existingAccountRequest_getsSuccessfully() throws InvalidParametersException {
        AccountRequest expectedAccountRequest =
                new AccountRequest("test@gmail.com", "name", "institute", AccountRequestStatus.PENDING, "comments");
        UUID id = expectedAccountRequest.getId();
        accountRequestsLogic.createAccountRequest(expectedAccountRequest);
        AccountRequest actualAccountRequest = accountRequestsLogic.getAccountRequest(id);
        assertEquals(expectedAccountRequest, actualAccountRequest);
    }

    @Test
    public void testResetAccountRequest()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        ______TS("success: create account request and update registeredAt field");

        String name = "name lee";
        String email = "email@gmail.com";
        String institute = "institute";
        AccountRequestStatus status = AccountRequestStatus.PENDING;
        String comments = "comments";

        AccountRequest toReset = accountRequestsLogic.createAccountRequest(name, email, institute, status, comments);
        AccountRequestsDb accountRequestsDb = AccountRequestsDb.inst();

        toReset.setRegisteredAt(Instant.now());
        UUID id = toReset.getId();
        toReset = accountRequestsDb.getAccountRequest(id);

        assertNotNull(toReset);
        assertNotNull(toReset.getRegisteredAt());

        ______TS("success: reset account request that already exists");

        AccountRequest resetted = accountRequestsLogic.resetAccountRequest(id);

        assertNull(resetted.getRegisteredAt());

        ______TS("success: test delete account request");

        accountRequestsLogic.deleteAccountRequest(toReset.getId());

        assertNull(accountRequestsLogic.getAccountRequest(toReset.getId()));

        ______TS("failure: reset account request that does not exist");

        assertThrows(EntityDoesNotExistException.class,
                () -> accountRequestsLogic.resetAccountRequest(id));
    }
}
