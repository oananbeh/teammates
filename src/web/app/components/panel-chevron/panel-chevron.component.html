<button type="button" class="chevron btn-course btn btn-primary btn-sm" aria-label="Expand panel" [attr.aria-expanded]=isExpanded>
  <i class="fas fa-chevron-down" [style.color]="chevronColor" *ngIf="!isExpanded"></i>
  <i class="fas fa-chevron-up" [style.color]="chevronColor" *ngIf="isExpanded"></i>
</button>
