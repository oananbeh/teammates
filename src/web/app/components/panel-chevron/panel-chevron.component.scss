button.chevron {
  display: inline-flex;
  vertical-align: middle;
  background-color: inherit;
  border-width: 0;
}

button.chevron:active {
  background-color: inherit;
}
