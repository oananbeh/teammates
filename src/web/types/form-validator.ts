import { ApiConst, ApiStringConst } from './api-const';

/**
 * Represents the root FormValidator object of all form fields.
 */
export enum FormValidator {
  /**
   * Max length for the 'Course ID' field.
   */
  COURSE_ID_MAX_LENGTH = ApiConst.COURSE_ID_MAX_LENGTH,

  /**
   * Max length for the 'Course Name' field.
   */
  COURSE_NAME_MAX_LENGTH = ApiConst.COURSE_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Student Name` field.
   */
  STUDENT_NAME_MAX_LENGTH = ApiConst.STUDENT_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Section Name` field.
   */
  SECTION_NAME_MAX_LENGTH = ApiConst.SECTION_NAME_MAX_LENGTH,

  /**
   * Max length for the 'Team Name` field.
   */
  TEAM_NAME_MAX_LENGTH = ApiConst.TEAM_NAME_MAX_LENGTH,

  /**
   * Max length for the 'E-mail Address` field.
   */
  EMAIL_MAX_LENGTH = ApiConst.EMAIL_MAX_LENGTH,

  /**
   * Regex used to verify emails in the back-end.
   */
  EMAIL_REGEX = ApiStringConst.EMAIL_REGEX,

  /**
   * Regex used to verify names.
   *
   * Based on back-end's `FieldValidator.REGEX_NAME`.
   * The back-end regex is not converted to use here as the pattern syntax is not accepted in JS.
   */
  NAME_REGEX = '^[a-zA-Z0-9][^|%]*$',

  /**
   * Regex used to verify country names.
   *
   * Based on back-end's `FieldValidator.REGEX_NAME`, but without needing to start with alphanumeric
   * as the country is added to the end of the combined institute string.
   */
  COUNTRY_REGEX = '^[^|%]*$',

  /**
   * Max length for institution name in account request. (to be combined with country)
   */
  INSTITUTION_NAME_MAX_LENGTH = 86,

  /**
   * Max length for country in account request. (to be combined with institution name)
   */
  COUNTRY_NAME_MAX_LENGTH = 40,
}
