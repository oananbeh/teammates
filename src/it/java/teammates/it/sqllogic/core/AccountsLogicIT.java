package teammates.it.sqllogic.core;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.NotificationStyle;
import teammates.common.datatransfer.NotificationTargetUser;
import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.exception.EntityAlreadyExistsException;
import teammates.common.exception.EntityDoesNotExistException;
import teammates.common.exception.InvalidParametersException;
import teammates.common.util.FieldValidator;
import teammates.common.util.HibernateUtil;
import teammates.common.util.StringHelper;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.AccountsLogic;
import teammates.sqllogic.core.CoursesLogic;
import teammates.sqllogic.core.NotificationsLogic;
import teammates.sqllogic.core.UsersLogic;
import teammates.storage.sqlapi.AccountsDb;
import teammates.storage.sqlentity.Account;
import teammates.storage.sqlentity.Course;
import teammates.storage.sqlentity.Instructor;
import teammates.storage.sqlentity.Notification;
import teammates.storage.sqlentity.ReadNotification;
import teammates.storage.sqlentity.Student;
import teammates.test.AssertHelper;

/**
 * SUT: {@link AccountsLogic}.
 */
public class AccountsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private AccountsLogic accountsLogic = AccountsLogic.inst();
    private NotificationsLogic notificationsLogic = NotificationsLogic.inst();
    private UsersLogic usersLogic = UsersLogic.inst();
    private CoursesLogic coursesLogic = CoursesLogic.inst();

    private AccountsDb accountsDb = AccountsDb.inst();

    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
        HibernateUtil.clearSession();
    }

    @Test
    public void testUpdateReadNotifications()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {
        ______TS("success: mark notification as read");
        Account account = new Account("google-id", "name", "email@teammates.com");
        Notification notification = new Notification(Instant.parse("2011-01-01T00:00:00Z"),
                Instant.parse("2099-01-01T00:00:00Z"), NotificationStyle.DANGER, NotificationTargetUser.GENERAL,
                "A deprecation note", "<p>Deprecation happens in three minutes</p>");
        accountsDb.createAccount(account);
        notificationsLogic.createNotification(notification);

        String googleId = account.getGoogleId();
        UUID notificationId = notification.getId();
        accountsLogic.updateReadNotifications(googleId, notificationId, notification.getEndTime());

        Account actualAccount = accountsDb.getAccountByGoogleId(googleId);
        List<ReadNotification> accountReadNotifications = actualAccount.getReadNotifications();
        assertEquals(1, accountReadNotifications.size());
        assertSame(actualAccount, accountReadNotifications.get(0).getAccount());
        assertSame(notification, accountReadNotifications.get(0).getNotification());
    }

    @Test
    public void testJoinCourseForStudent()
            throws EntityAlreadyExistsException, InvalidParametersException, EntityDoesNotExistException {

        Student student2YetToJoinCourse = typicalDataBundle.students.get("student2YetToJoinCourse4");
        Student student3YetToJoinCourse = typicalDataBundle.students.get("student3YetToJoinCourse4");
        Student studentInCourse = typicalDataBundle.students.get("student1InCourse1");

        String loggedInGoogleId = "AccLogicT.student.id";

        ______TS("failure: wrong key");

        String wrongKey = StringHelper.encrypt("wrongkey");
        EntityDoesNotExistException ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForStudent(wrongKey, loggedInGoogleId));
        assertEquals("No student with given registration key: " + wrongKey, ednee.getMessage());

        ______TS("failure: invalid parameters");

        InvalidParametersException ipe = assertThrows(InvalidParametersException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(), "wrong student"));
        AssertHelper.assertContains(FieldValidator.REASON_INCORRECT_FORMAT, ipe.getMessage());

        ______TS("failure: googleID belongs to an existing student in the course");

        EntityAlreadyExistsException eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(),
                studentInCourse.getGoogleId()));
        assertEquals("Student has already joined course", eaee.getMessage());

        ______TS("success: with encryption and new account to be created");

        accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(), loggedInGoogleId);
        Account accountCreated = accountsLogic.getAccountForGoogleId(loggedInGoogleId);

        assertEquals(loggedInGoogleId, usersLogic.getStudentForEmail(
                student2YetToJoinCourse.getCourseId(), student2YetToJoinCourse.getEmail()).getGoogleId());
        assertNotNull(accountCreated);

        ______TS("success: student joined but account already exists");

        String existingAccountId = "existingAccountId";
        Account existingAccount = new Account(existingAccountId, "accountName", student3YetToJoinCourse.getEmail());
        accountsDb.createAccount(existingAccount);

        accountsLogic.joinCourseForStudent(student3YetToJoinCourse.getRegKey(), existingAccountId);

        assertEquals(existingAccountId, usersLogic.getStudentForEmail(
                student3YetToJoinCourse.getCourseId(), student3YetToJoinCourse.getEmail()).getGoogleId());

        ______TS("failure: already joined");

        eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(), loggedInGoogleId));
        assertEquals("Student has already joined course", eaee.getMessage());

        ______TS("failure: course is deleted");

        Course originalCourse = usersLogic.getStudentForEmail(
                student2YetToJoinCourse.getCourseId(), student2YetToJoinCourse.getEmail()).getCourse();
        coursesLogic.moveCourseToRecycleBin(originalCourse.getId());

        ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForStudent(student2YetToJoinCourse.getRegKey(),
                        loggedInGoogleId));
        assertEquals("The course you are trying to join has been deleted by an instructor", ednee.getMessage());
    }

    @Test
    public void testJoinCourseForInstructor() throws Exception {
        String instructorIdAlreadyJoinedCourse = "instructor1";
        Instructor instructor2YetToJoinCourse = typicalDataBundle.instructors.get("instructor2YetToJoinCourse4");
        Instructor instructor3YetToJoinCourse = typicalDataBundle.instructors.get("instructor3YetToJoinCourse4");

        String loggedInGoogleId = "AccLogicT.instr.id";
        String[] key = new String[] {
                getRegKeyForInstructor(instructor2YetToJoinCourse.getCourseId(), instructor2YetToJoinCourse.getEmail()),
                getRegKeyForInstructor(instructor2YetToJoinCourse.getCourseId(), instructor3YetToJoinCourse.getEmail()),
        };

        ______TS("failure: googleID belongs to an existing instructor in the course");

        EntityAlreadyExistsException eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForInstructor(
                        key[0], instructorIdAlreadyJoinedCourse));
        assertEquals("Instructor has already joined course", eaee.getMessage());

        ______TS("success: instructor joined and new account be created");

        accountsLogic.joinCourseForInstructor(key[0], loggedInGoogleId);

        Instructor joinedInstructor = usersLogic.getInstructorForEmail(
                        instructor2YetToJoinCourse.getCourseId(), instructor2YetToJoinCourse.getEmail());
        assertEquals(loggedInGoogleId, joinedInstructor.getGoogleId());

        Account accountCreated = accountsLogic.getAccountForGoogleId(loggedInGoogleId);
        assertNotNull(accountCreated);

        ______TS("success: instructor joined but account already exists");

        String existingAccountId = "existingAccountId";
        Account existingAccount = new Account(existingAccountId, "accountName", instructor3YetToJoinCourse.getEmail());
        accountsDb.createAccount(existingAccount);

        accountsLogic.joinCourseForInstructor(key[1], existingAccount.getGoogleId());

        joinedInstructor = usersLogic.getInstructorForEmail(
                        instructor3YetToJoinCourse.getCourseId(), existingAccount.getEmail());
        assertEquals(existingAccountId, joinedInstructor.getGoogleId());

        ______TS("failure: instructor already joined");

        eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForInstructor(key[0], loggedInGoogleId));
        assertEquals("Instructor has already joined course", eaee.getMessage());

        ______TS("failure: key belongs to a different user");

        eaee = assertThrows(EntityAlreadyExistsException.class,
                () -> accountsLogic.joinCourseForInstructor(key[0], "otherUserId"));
        assertEquals("Instructor has already joined course", eaee.getMessage());

        ______TS("failure: invalid key");

        String invalidKey = StringHelper.encrypt("invalidKey");

        EntityDoesNotExistException ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForInstructor(invalidKey, loggedInGoogleId));
        assertEquals("No instructor with given registration key: " + invalidKey,
                ednee.getMessage());

        ______TS("failure: course deleted");

        Course originalCourse = usersLogic.getInstructorForEmail(
                instructor2YetToJoinCourse.getCourseId(), instructor2YetToJoinCourse.getEmail()).getCourse();
        coursesLogic.moveCourseToRecycleBin(originalCourse.getId());

        ednee = assertThrows(EntityDoesNotExistException.class,
                () -> accountsLogic.joinCourseForInstructor(instructor2YetToJoinCourse.getRegKey(),
                    instructor2YetToJoinCourse.getGoogleId()));
        assertEquals("The course you are trying to join has been deleted by an instructor", ednee.getMessage());
    }

    private String getRegKeyForInstructor(String courseId, String email) {
        return usersLogic.getInstructorForEmail(courseId, email).getRegKey();
    }
}
