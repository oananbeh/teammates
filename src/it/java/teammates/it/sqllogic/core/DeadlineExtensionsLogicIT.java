package teammates.it.sqllogic.core;

import java.time.Instant;

import org.testng.annotations.BeforeClass;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import teammates.common.datatransfer.SqlDataBundle;
import teammates.common.util.HibernateUtil;
import teammates.it.test.BaseTestCaseWithSqlDatabaseAccess;
import teammates.sqllogic.core.DeadlineExtensionsLogic;
import teammates.storage.sqlentity.FeedbackSession;
import teammates.storage.sqlentity.Student;

/**
 * SUT: {@link DeadlineExtensionsLogic}.
 */
public class DeadlineExtensionsLogicIT extends BaseTestCaseWithSqlDatabaseAccess {

    private DeadlineExtensionsLogic deadlineExtensionsLogic = DeadlineExtensionsLogic.inst();
    private SqlDataBundle typicalDataBundle;

    @Override
    @BeforeClass
    public void setupClass() {
        super.setupClass();
        typicalDataBundle = getTypicalSqlDataBundle();
    }

    @Override
    @BeforeMethod
    protected void setUp() throws Exception {
        super.setUp();
        persistDataBundle(typicalDataBundle);
        HibernateUtil.flushSession();
        HibernateUtil.clearSession();
    }

    @Test
    public void testGetExtendedDeadline_extensionExists_success() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student1InCourse1");

        assert student != null;
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNotNull(extendedDeadlineForStudent);
        assertEquals(Instant.parse("2027-04-30T23:00:00Z"), extendedDeadlineForStudent);
    }

    @Test
    public void testGetExtendedDeadline_extensionDoesNotExist_null() {
        FeedbackSession feedbackSession = typicalDataBundle.feedbackSessions.get("session1InCourse1");
        Student student = typicalDataBundle.students.get("student2InCourse1");
        Instant extendedDeadlineForStudent = deadlineExtensionsLogic.getExtendedDeadlineForUser(feedbackSession, student);

        assertNull(extendedDeadlineForStudent);
    }
}
