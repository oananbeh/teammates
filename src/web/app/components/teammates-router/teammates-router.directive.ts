import { LocationStrategy } from '@angular/common';
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { ActivatedRoute, Router, RouterLink } from '@angular/router';
import { MasqueradeModeService } from '../../../services/masquerade-mode.service';

/**
 * Router link that preserves masquerade mode
 */
@Directive({
  selector: 'a[tmRouterLink]',
})
export class TeammatesRouterDirective extends RouterLink {
  private queryParamsInternal: { [k: string]: any } = {};

  @Input()
  set tmRouterLink(commands: any[] | string) {
    this.routerLink = commands;
  }

  @Input()
  // @ts-expect-error query params is redefined in this class
  set queryParams(params: { [k: string]: any }) {
    this.queryParamsInternal = params;
  }

  override get queryParams(): { [k: string]: any } {
    const userParam: string = this.masqueradeModeService.getMasqueradeUser();
    if (userParam !== '') {
      return { ...this.queryParamsInternal, user: userParam };
    }
    return this.queryParamsInternal;
  }

  constructor(router: Router, route: ActivatedRoute,
              renderer: Renderer2, el: ElementRef, locationStrategy: LocationStrategy,
              private masqueradeModeService: MasqueradeModeService) {
    super(router, route, null, renderer, el, locationStrategy);
  }
}
