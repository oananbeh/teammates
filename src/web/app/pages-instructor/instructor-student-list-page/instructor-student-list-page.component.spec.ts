import { HttpStatusCode } from '@angular/common/http';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { throwError } from 'rxjs';
import { CourseTab, InstructorStudentListPageComponent } from './instructor-student-list-page.component';
import { InstructorStudentListPageModule } from './instructor-student-list-page.module';
import { StudentService } from '../../../services/student.service';
import { Course } from '../../../types/api-output';
import { SortBy, SortOrder } from '../../../types/sort-properties';
import { PanelChevronModule } from '../../components/panel-chevron/panel-chevron.module';
import { TeammatesRouterModule } from '../../components/teammates-router/teammates-router.module';

describe('InstructorStudentListPageComponent', () => {
  let component: InstructorStudentListPageComponent;
  let fixture: ComponentFixture<InstructorStudentListPageComponent>;
  let studentService: StudentService;

  const course1: Course = {
    courseId: 'course1Id',
    courseName: 'Course 1',
    timeZone: 'UTC',
    institute: 'Institute',
    creationTimestamp: 1649791778732,
    deletionTimestamp: 0,
  };

  const course1Tab: CourseTab = {
    course: course1,
    studentList: [
      {
        student: {
          email: 'student1@example.com',
          courseId: 'course1Id',
          name: 'Student 1',
          teamName: 'Team 1',
          sectionName: 'Section 1',
        },
        isAllowedToViewStudentInSection: true,
        isAllowedToModifyStudent: true,
      },
      {
        student: {
          email: 'student2@example.com',
          courseId: 'course1Id',
          name: 'Student 2',
          teamName: 'Team 1',
          sectionName: 'Section 1',
        },
        isAllowedToViewStudentInSection: true,
        isAllowedToModifyStudent: true,
      },
      {
        student: {
          email: 'student3@example.com',
          courseId: 'course1Id',
          name: 'Student 3',
          teamName: 'Team 4',
          sectionName: 'Section 5',
        },
        isAllowedToViewStudentInSection: true,
        isAllowedToModifyStudent: true,
      },
    ],
    studentSortBy: SortBy.NONE,
    studentSortOrder: SortOrder.ASC,
    hasTabExpanded: false,
    hasStudentLoaded: false,
    hasLoadingFailed: false,
    isAbleToViewStudents: true,
    stats: {
      numOfSections: 0,
      numOfStudents: 0,
      numOfTeams: 0,
    },
  };

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      imports: [
        HttpClientTestingModule,
        TeammatesRouterModule,
        RouterTestingModule,
        FormsModule,
        InstructorStudentListPageModule,
        PanelChevronModule,
        BrowserAnimationsModule,
      ],
      providers: [
        StudentService,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(InstructorStudentListPageComponent);
    studentService = TestBed.inject(StudentService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should block instructors from viewing student details if they do not have the permission', () => {
    jest.spyOn(studentService, 'getStudentsFromCourse').mockReturnValue(throwError(() => ({
      status: HttpStatusCode.Forbidden,
      error: {
        message: 'You are not authorized to access this resource.',
      },
    })));
    component.loadStudents(course1Tab);
    expect(course1Tab.isAbleToViewStudents).toBeFalsy();
    expect(course1Tab.hasStudentLoaded).toBeTruthy();
    expect(course1Tab.studentList.length).toEqual(0);
  });

  it('should snap with a course with students the instructor has no permission to view', () => {
    component.courseTabList.push(course1Tab);
    component.isLoadingCourses = false;
    component.courseTabList[0].hasTabExpanded = true;
    component.courseTabList[0].hasStudentLoaded = true;
    component.courseTabList[0].isAbleToViewStudents = false;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
