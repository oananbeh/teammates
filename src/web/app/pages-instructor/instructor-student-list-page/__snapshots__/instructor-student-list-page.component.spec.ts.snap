// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstructorStudentListPageComponent should snap with a course with students the instructor has no permission to view 1`] = `
<tm-instructor-student-list-page
  SortBy={[Function Object]}
  courseService={[Function CourseService]}
  courseTabList={[Function Array]}
  coursesSortBy={[Function Number]}
  hasLoadingFailed="false"
  instructorService={[Function InstructorService]}
  isLoadingCourses="false"
  statusMessageService={[Function StatusMessageService]}
  studentService={[Function StudentService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <h1>
    Students
  </h1><p>
     This page gives the student details for all your courses. Expand the panels below to view the details of students in each course. Use the 
    <a
      href="/web/instructor/search"
      tmrouterlink="/web/instructor/search"
    >
      'Search' page
    </a>
     if you need to locate the details of a specific student by name or email.
  </p><hr /><tm-loading-retry>
    <div />
    <div
      class="row mb-3"
    >
      <div
        class="col"
      >
        <div
          class="float-end"
        >
          <strong
            class="d-inline"
          >
             Sort By: 
          </strong>
          <div
            class="btn-group"
            data-toggle="buttons"
          >
            <button
              class="btn btn-light"
              id="sort-course-id"
            >
              Course ID
            </button>
            <button
              class="btn btn-light"
              id="sort-course-name"
            >
              Course Name
            </button>
            <button
              class="btn btn-light"
              disabled=""
              id="sort-course-date"
            >
              Creation Date
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      class="course-table"
    >
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [course1Id]: Course 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div
          class="ng-trigger ng-trigger-collapseAnim ng-animating"
        >
          <tm-loading-retry>
            <div
              style=""
            >
              <div
                class="card-body"
              >
                <div
                  class="alert alert-danger"
                  role="alert"
                >
                  <h5>
                    You do not have permission to view the details of the students in this course.
                  </h5>
                </div>
              </div>
            </div>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </tm-loading-retry>
</tm-instructor-student-list-page>
`;
